lib/Refactoring/SyntacticRenameRangeDetails.cpp

```diff
@@ -366,89 +364,153 @@ RegionType RenameRangeDetailCollector::getSyntacticRenameRegionType(
+    // base name, which does not match. Subscripts are never called in the
+    // index.
+    if (usage == RenameLocUsage::Definition) {
+      if (renameBase(resolved.range, RefactoringRangeKind::KeywordBaseName)) {
+        return RegionType::Mismatch;
+      }
+    }
+    break;
   }
 
-  bool HandleLabels = false;
-  if (Config.IsFunctionLike) {
-    switch (Config.Usage) {
-    case NameUsage::Call:
-      HandleLabels = !isOperator();
+  bool handleLabels = false;
+  bool isCallSite = false;
+  if (Old.isFunction()) {
+    switch (usage) {
+    case RenameLocUsage::Call:
+      // All calls except for operators have argument labels that should be
+      // renamed.
+      handleLabels = !Lexer::isOperator(Old.base());
+      isCallSite = true;
       break;
-    case NameUsage::Definition:
-      HandleLabels = true;
+    case RenameLocUsage::Definition:
+      // All function definitions have argument labels that should be renamed.
+      handleLabels = true;
+      isCallSite = false;
       break;
-    case NameUsage::Reference:
-      HandleLabels =
-          Resolved.labelType == LabelRangeType::Selector || IsSubscript;
+    case RenameLocUsage::Reference:
+      if (resolved.labelType == LabelRangeType::CompoundName) {
+        // If we have a compound name that specifies argument labels to
+        // disambiguate functions with the same base name, we always need to
+        // rename the labels.
+        handleLabels = true;
+        isCallSite = false;
+      } else if (specialBaseName == SpecialBaseName::Subscript) {
+        // Accesses to a subscript are modeled using a reference to the
+        // subscript (with base name `[`). We always need to rename argument
+        // labels here.
+        handleLabels = true;
+        isCallSite = true;
+      } else {
+        handleLabels = false;
+        isCallSite = false;
+      }
       break;
-    case NameUsage::Unknown:
-      HandleLabels = Resolved.labelType != LabelRangeType::None;
+    case RenameLocUsage::Unknown:
+      // If we don't know where the function is used, fall back to trying to
+      // rename labels if there are some.
+      handleLabels = resolved.labelType != LabelRangeType::None;
+      isCallSite = resolved.labelType == LabelRangeType::CallArg;
       break;
     }
   }
 
-  if (HandleLabels) {
-    bool isCallSite = Config.Usage != NameUsage::Definition &&
-                      (Config.Usage != NameUsage::Reference || IsSubscript) &&
-                      Resolved.labelType == LabelRangeType::CallArg;
-
-    if (renameLabels(Resolved.labelRanges, Resolved.firstTrailingLabel,
-                     Resolved.labelType, isCallSite))
-      return Config.Usage == NameUsage::Unknown ? RegionType::Unmatched
-                                                : RegionType::Mismatch;
+  if (!handleLabels) {
+    return regionKind;
+  }
+  if (renameLabels(resolved.labelRanges, resolved.firstTrailingLabel,
```

Comment:
```suggestion
  if (handleLabels && renameLabels(resolved.labelRanges, resolved.firstTrailingLabel,
```

Reply:
I prefer my version. `renameLabels` is already such a long call that I don’t want to make the condition even longer, and thus less readable IMO.